src/goto-programs/goto_convert.cpp

```diff
@@ -718,16 +718,14 @@ void goto_convertt::convert_assign(
 
     if(lhs.id() == ID_typecast)
     {
-      DATA_INVARIANT(
-        lhs.operands().size() == 1, "Typecast must have one operand");
+      UNREACHABLE; // the original code here was broken, presumably it's never used
+      typecast_exprt tc = to_typecast_expr(lhs);
```

Comment:
I think some compilers will give you a hard time for this unreachable code

Reply:
I have no idea how we would generate an expression that has a typecast on the left-hand side, but even if and when we do: at least goto-symex does handle that case, so we could safely leave it in the goto program. I am going to add a commit that simply removes this code.

Reply:
We definitely do sometimes produce `void** data = new void*[10]; ((Object**)data)[0] = new Object` in the Java frontend. Whether we should is another question!

Reply:
Sure, and you can make something similar happen in C (`*(char*)&x = 42;`) - but there's an `index_exprt` on the left-hand side in your example and a `dereference_exprt` in my case. Neither has a typecast at the top level on the left-hand side.

Reply:
True. In the past we would sometimes do `(Object*)var = new Object`, but AFAIK that's no longer generated by the frontend

Reply:
I think even if that were still the case, goto-symex would happily take care of that.

Reply:
Just for the record, those changes have now been moved to #3770, this PR now exclusively deals with replacing `make_typecast`.